components/wallet/nwc.js

```diff
@@ -0,0 +1,161 @@
+    type: 'password'
+  }
+]
+
+export const card = {
+  title: 'NWC',
+  subtitle: 'use Nostr Wallet Connect for payments',
+  badges: ['send only', 'non-custodialish']
+}
+
+export const schema = object({
+  nwcUrl: string()
+    .required('required')
+    .test(async (nwcUrl, context) => {
+      // run validation in sequence to control order of errors
+      // inspired by https://github.com/jquense/yup/issues/851#issuecomment-1049705180
+      try {
+        await string().required('required').validate(nwcUrl)
+        await string().matches(/^nostr\+?walletconnect:\/\//, { message: 'must start with nostr+walletconnect://' }).validate(nwcUrl)
+        let relayUrl, walletPubkey, secret
+        try {
+          ({ relayUrl, walletPubkey, secret } = parseNwcUrl(nwcUrl))
+        } catch {
+          // invalid URL error. handle as if pubkey validation failed to not confuse user.
+          throw new Error('pubkey must be 64 hex chars')
+        }
+        await string().required('pubkey required').trim().matches(NOSTR_PUBKEY_HEX, 'pubkey must be 64 hex chars').validate(walletPubkey)
+        await string().required('relay url required').trim().wss('relay must use wss://').validate(relayUrl)
+        await string().required('secret required').trim().matches(/^[0-9a-fA-F]{64}$/, 'secret must be 64 hex chars').validate(secret)
+      } catch (err) {
+        return context.createError({ message: err.message })
+      }
+      return true
+    })
+})
+
+export async function validate ({ logger, nwcUrl }) {
+  const { relayUrl, walletPubkey } = parseNwcUrl(nwcUrl)
+
+  logger.info(`requesting info event from ${relayUrl}`)
+  const relay = await Relay
+    .connect(relayUrl)
+    .catch(() => {
+      // NOTE: passed error is undefined for some reason
+      const msg = `failed to connect to ${relayUrl}`
+      logger.error(msg)
+      throw new Error(msg)
+    })
+  logger.ok(`connected to ${relayUrl}`)
+
+  try {
+    return await new Promise((resolve, reject) => {
+      let found = false
+      const sub = relay.subscribe([
+        {
+          kinds: [13194],
+          authors: [walletPubkey]
+        }
+      ], {
+        onevent (event) {
+          found = true
+          logger.ok(`received info event from ${relayUrl}`)
+          resolve(event)
+        },
+        onclose (reason) {
+          if (!['closed by caller', 'relay connection closed by us'].includes(reason)) {
+            // only log if not closed by us (caller)
+            const msg = 'connection closed: ' + (reason || 'unknown reason')
+            logger.error(msg)
+            reject(new Error(msg))
+          }
+        },
+        oneose () {
+          if (!found) {
+            const msg = 'EOSE received without info event'
+            logger.error(msg)
+            reject(new Error(msg))
+          }
+          sub?.close()
+        }
+      })
+    })
+  } finally {
+    // For some reason, websocket is already in CLOSING or CLOSED state.
```

Comment:
probably because it's garbage collected once the `try` block is finished

Reply:
Mhh, that `sub?.close()` should only close the subscription. But this comment is about the relay connection itself. Just because a subscription was closed, the relay connection shouldn't close? Mhh, but maybe it's closing the relay because it's the last subscription that was closed? I need to check the library code. 

This is also a comment from when I implemented NWC.

Reply:
I'd guess it doesn't keep the connection open without an active subscription.